packages/core/src/codewhisperer/tracker/userWrittenCodeTracker.ts

```diff
@@ -0,0 +1,196 @@
+/*!
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { getLogger } from '../../shared/logger/logger'
+import { runtimeLanguageContext } from '../util/runtimeLanguageContext'
+import { TelemetryHelper } from '../util/telemetryHelper'
+import { AuthUtil } from '../util/authUtil'
+import { getSelectedCustomization } from '../util/customizationUtil'
+import { codeWhispererClient as client } from '../client/codewhisperer'
+import { isAwsError } from '../../shared/errors'
+import { CodewhispererLanguage } from '../../shared'
+
+/**
+ * This singleton class is mainly used for calculating the user written code
+ * for active Amazon Q users.
+ * It reports the user written code per 5 minutes when the user is coding and using Amazon Q features
+ */
+export class UserWrittenCodeTracker {
+    private _userWrittenNewCodeCharacterCount: Map<CodewhispererLanguage, number>
+    private _userWrittenNewCodeLineCount: Map<CodewhispererLanguage, number>
+    private _qIsMakingEdits: boolean
+    private _timer?: NodeJS.Timer
+    private _qUsageCount: number
+    private _lastQInvocationTime: number
+
+    static #instance: UserWrittenCodeTracker
+    static copySnippetThreshold = 50
+    static resetQIsEditingTimeoutMs = 2 * 60 * 1000
+    static defaultCheckPeriodMillis = 5 * 60 * 1000
+
+    private constructor() {
+        this._userWrittenNewCodeLineCount = new Map<CodewhispererLanguage, number>()
+        this._userWrittenNewCodeCharacterCount = new Map<CodewhispererLanguage, number>()
+        this._qUsageCount = 0
+        this._qIsMakingEdits = false
+        this._timer = undefined
+        this._lastQInvocationTime = 0
+    }
+
+    public static get instance() {
+        return (this.#instance ??= new this())
+    }
+
+    public isActive(): boolean {
+        return TelemetryHelper.instance.isTelemetryEnabled() && AuthUtil.instance.isConnected()
+    }
+
+    // this should be invoked whenever there is a successful Q feature invocation
+    // for all Q features
+    public onQFeatureInvoked() {
+        this._qUsageCount += 1
+        this._lastQInvocationTime = performance.now()
+    }
+
+    public onQStartsMakingEdits() {
+        this._qIsMakingEdits = true
+    }
+
+    public onQFinishesEdits() {
+        this._qIsMakingEdits = false
+    }
```

Comment:
thinking of a case where there are multiple Q features are working at the same time, how do we deal with this

for example, while `transformation` is working and users trigger a inline completion request

Reply:
This is Q working flag only controls the moment when Q feature calls the editor edit api, it is almost instant.

Reply:
It is not the time when transformation is pending

Reply:
isee